src-exe/cabal-plan.hs

```diff
@@ -868,16 +869,99 @@ trPairs :: Tr a -> [(Int,a,a)]
 trPairs (No _ i js) =
     [ (n, i, j) | No n j _ <- js ] ++ concatMap trPairs js
 
+-- | Calculate the set of nodes reachable from a given node (including the node
+-- itself).
+reachableFrom :: (Ord i, Ord v) => TG.G v i -> v -> Set v
+reachableFrom graph =
+    \v -> S.insert v (M.findWithDefault S.empty v reachable)
+  where
+    reachable = TG.adjacencyMap (TG.closure graph)
+
 doDot
     :: Flag ShowBuiltin
     -> Flag ShowGlobal
     -> PlanJson
     -> Flag DotTred
     -> Flag DotTredWght
-    -> [Highlight] -> IO ()
-doDot showBuiltin showGlobal plan tred tredWeights highlights = either loopGraph id $ TG.runG am $ \g' -> do
+    -> [Highlight]
+    -> [Pattern] -> IO ()
+doDot showBuiltin showGlobal plan tred tredWeights highlights rootPatterns = either loopGraph id $ TG.runG am $ \g' -> do
     let g = if fromFlag DotTred tred then TG.reduction g' else g'
 
+    let isReachableUnit :: DotUnitId -> Bool
+        isReachableUnit unitId =
+          null rootPatterns || S.member unitId reachableUnits
+          where
+            -- Units reachable from any unit matching any pattern given
+            reachableUnits :: Set DotUnitId
+            reachableUnits =
+              foldMap
+                (reachableFrom g)
+                rootUnits
+              where
+                rootUnits :: Set DotUnitId
+                rootUnits =
+                  S.filter
+                    (getAny . foldMap checkPatternDotUnit rootPatterns)
+                    dotUnits
+
+    let duShow :: DotUnitId -> Bool
+        duShow dotUnitId@(DU unitId _) = case M.lookup unitId units of
+          Nothing -> False
+          Just unit ->
+              if isReachableUnit dotUnitId
+              then case uType unit of
+                UnitTypeBuiltin -> fromFlag ShowBuiltin showBuiltin
+                UnitTypeGlobal  -> fromFlag ShowGlobal showGlobal
+                UnitTypeLocal   -> True
+                UnitTypeInplace -> True
+              else False
+
+    let vertex :: Set DotUnitId -> DotUnitId -> IO ()
+        vertex redVertices du = when (duShow du) $ T.putStrLn $ mconcat
+            [ "\""
+            , dispDotUnit du
+            , "\""
+            -- shape
+            , " [shape="
+            , duShape du
+            -- color
+            , ",color="
+            , color
+            , "];"
+            ]
+          where
+            color | S.member du redVertices = "red"
+                  | otherwise               = borderColor du
+
+    let edge
```

Comment:
why this are moved?

Reply:
They close over the set or reachable units, which is not available from the top level args - need `g` computed from `TG.runG`

Reply:
I see, that's OK.